File: cs50/pset2/vigenere.c
/****************************************************************************
 
* vigenere.c
*
 
* Provotorov Daniil
* provotorov.d.i@gmail.com
*
 
* Cipher text via vigenere cipher
*
 
***************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <cs50.h>
#include <string.h>
#include <ctype.h>

int get_key(string word, int counter);// прототип функцииc

int main(int argc, string argv[])
{
    if (argc != 2) // проверка на 2 агрумента командной строки
    {
        printf("Введите одно значение ключа\n"); // если аргументов не 2 - вернули ошибку
        return 1;
    }
    string key = argv[1];  // получили ключ

    for (int i = 0, n = strlen(key); i < n; i++)
    {
        if (isalpha(key[i]))
        {
        }
        else
        {
            printf("Введите корректное значение ключа\n");
            return 1;
        }
    }

    printf("plaintext: ");
    string plaintext = get_string(""); // запросили строку текста
    printf("ciphertext: ");

    for (int i = 0, n = strlen(plaintext), j = 0; i < n; i ++) //проходимся по тексту
    {
        if (isupper(plaintext[i])) // что делать, если верхняя буква
        {
            int alphabetic_index0 = plaintext[i] - 65;
            int alphabetic_index1 = (alphabetic_index0 + get_key(key, j)) % 26;
            j++;
            int cipherletter = 65 + alphabetic_index1;
            printf("%c", (char) cipherletter);
        }
        else if (islower(plaintext[i])) // что делать, если нижняя буква
        {
            int alphabetic_index0 = plaintext[i] - 97;
            int alphabetic_index1 = (alphabetic_index0 + get_key(key, j)) % 26;
            j++;
            int cipherletter = 97 + alphabetic_index1;
            printf("%c", (char) cipherletter);
        }
        else
        {
            printf("%c", (char) plaintext[i]);
        }
    }

    printf("\n");
    return 0;
}


// число, на которое сдвигаем нашу букву - kj
int get_key(string word, int counter)
{
    int n = strlen(word);
    int index = counter % n;
    if (isupper(word[index]))
    {
        return (word[index] - 65);
    }
    else if (islower(word[index]))
    {
        return (word[index] - 97);
    }
    else
    {
        printf("Ошибочный ключ\n");
        return 1;
    }
}

File: cs50/pset6/vigenere.py
/****************************************************************************
 
* vigenere.py
*
 
* Provotorov Daniil
* provotorov.d.i@gmail.com
*
 
* Cipher text via vigenere cipher
*
 
***************************************************************************/

# импортировали аргументы командной строки
from sys import argv
# импортировали get_string
from cs50 import get_string
# проверили на число аргументов
if len(argv) != 2:
    print('Enter only 2 command-line arguments')
    exit(1)

key = argv[1]
# проверка на буквенную строку
for c in key:
    if not c.isalpha():
        print('Key should consists only alphabetic characters')
        exit(1)

plaintext = get_string('plaintext: ')
print('ciphertext: ', end="")

# получаем число, на которую сдвигаем букву
def get_key(word, counter):
    n = len(word)
    index = counter % n
    if word[index].isupper():
        return (ord(word[index]) - 65)
    elif word[index].islower():
        return (ord(word[index]) - 97)
    else:
        print('Ошибочный ключ')
        exit(1)

j = 0
for i in range (len(plaintext)):
    if plaintext[i].isupper():
        ai0 = ord(plaintext[i]) - 65
        ai1 = (ai0 + get_key(key,j)) % 26
        j = j+1
        ci = 65 + ai1
        print(chr(ci), end="")
    elif plaintext[i].islower():
        ai0 = ord(plaintext[i]) - 97
        ai1 = (ai0 + get_key(key,j)) % 26
        j = j+1
        ci = 97 + ai1
        print(chr(ci), end="")
    else:
        print(plaintext[i], end="")
print()
exit(0)
